packages/@sanity/cli/test/shared/devServer.ts

```diff
@@ -0,0 +1,95 @@
+import {spawn} from 'child_process'
+import {cliBinPath, sanityEnv} from './environment'
+import {request, ResponseData} from './request'
+
+export function testStartCommand({
+  port,
+  cwd,
+  expectedTitle,
+}: {
+  port: number
+  cwd: string
+  expectedTitle: string
+}): Promise<string> {
+  return new Promise(async (resolve, reject) => {
+    const maxWaitForServer = 120000
+    const startedAt = Date.now()
+    let hasSucceeded = false
+    let timer: ReturnType<typeof setTimeout>
+
+    const connect = await request(`http://localhost:${port}/`).catch((err) => err as Error)
+    if ('statusCode' in connect) {
+      reject(new Error(`Something is already listening on :${port}`))
+      return
+    }
+
+    const proc = spawn(process.argv[0], [cliBinPath, 'start'], {
+      cwd,
+      env: sanityEnv,
+      stdio: 'pipe',
+    })
+
+    const stderr: Buffer[] = []
+    const stdout: Buffer[] = []
+    proc.stderr.on('data', (chunk) => stderr.push(chunk))
+    proc.stdout.on('data', (chunk) => stdout.push(chunk))
+
+    proc.on('close', (code) => {
+      if (!hasSucceeded && code && code > 0) {
+        const stderrStr = Buffer.concat(stderr).toString('utf8')
+        const stdoutStr = Buffer.concat(stdout).toString('utf8')
+        reject(new Error(`'sanity start' failed with code ${code}:\n${stderrStr}\n${stdoutStr}`))
+      }
+    })
+
+    scheduleConnect()
+
+    function scheduleConnect() {
+      if (timer) {
+        clearTimeout(timer)
+      }
+
+      if (Date.now() - startedAt > maxWaitForServer) {
+        reject(new Error('Timed out waiting for server to get online'))
+        return
+      }
+
+      timer = setTimeout(tryConnect, 1000)
+    }
+
+    async function tryConnect() {
+      let res: ResponseData
+      try {
+        res = await Promise.race([
+          request(`http://localhost:${port}/`),
+          new Promise<ResponseData>((_, rejectTimeout) => setTimeout(rejectTimeout, 500)),
```

Comment:
Should `rejectTimeout` be called with an `Error` instance here?

Reply:
I suppose. We don't use the error for anything - just schedule another connection attempt - but looks a little cleaner with an error I suppose.